subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/projectmodule/DefaultProjectDependencyPublicationResolver.java

```diff
@@ -93,10 +124,41 @@ private <T> T resolve(Class<T> coordsType, Path identityPath, ProjectInternal pr
         }
 
         if (topLevelWithComponent.size() == 1) {
-            return topLevelWithComponent.iterator().next().getCoordinates(coordsType);
+            SoftwareComponentInternal singleComponent = topLevelWithComponent.iterator().next().getComponent().get();
+            Map<SoftwareComponent, T> componentCoordinates = getComponentCoordinates(coordsType, publications.keySet());
+            return new MultiCoordinateVariantResolver<>(singleComponent, identityPath, componentCoordinates);
         }
 
         // See if all entry points have the same identifier
+        return VariantCoordinateResolver.of(getCommonCoordinates(project, coordsType, topLevel));
+    }
+
+    /**
+     * Get the coordinates of a project that has no publications.
+     */
+    private static <T> T getImplicitCoordinates(Class<T> coordsType, Project project) {
+        // Project has no publications: simply use the project name in place of the dependency name
+        if (coordsType.isAssignableFrom(ModuleVersionIdentifier.class)) {
+
+            // TODO: Deprecate this behavior
+//        String message = "Cannot publish dependency on " + project.getDisplayName() + " since it does not declare any publications. " +
+//            "Publishing a component that depends on another project without publications";
+//        DeprecationLogger.deprecate(message)
+//            .withAdvice("Ensure " + project.getDisplayName() + " declares at least one publication.")
+//            .willBecomeAnErrorInGradle9()
+//            .withUpgradeGuideSection(8, "publishing_dependency_on_unpublished_project")
+//            .nagUser();
+
+            return coordsType.cast(DefaultModuleVersionIdentifier.newId(project.getGroup().toString(), project.getName(), project.getVersion().toString()));
+        }
+
+        throw new UnsupportedOperationException(String.format("Could not find any publications of type %s in %s.", coordsType.getSimpleName(), project.getDisplayName()));
```

Comment:
IMO the odd thing is the behavior above this, I think the message is fine. 

If a user has a dependency on a project, and that project doesn't publish anything in the coordinate system you are working in, we should say something like

"The project you're depending on does not publish anything with the coordinates you are asking for"

Which is close (ish) to the message there. 

But, we have this weird edge case where if you are looking for `ModuleVersionIdentifier`, we pretend that the project publishes something. This is wrong since once you try to consume the published components, that target project will not be there. 

This was an existing message though so I didn't update the wording.